src/Cimpress.Extensions.Http.Caching.InMemory/InMemoryCacheFallbackHandler.cs

```diff
@@ -0,0 +1,136 @@
+        /// <param name="cacheDuration">The maximum time span the item should be remained in the cache if not renewed before then.</param>
+        /// <param name="statsProvider">An <see cref="IStatsProvider"/> that records statistic information about the caching behavior.</param>
+        public InMemoryCacheFallbackHandler(HttpMessageHandler innerHandler, TimeSpan maxTimeout, TimeSpan cacheDuration, IStatsProvider statsProvider = null)
+            : this(innerHandler, maxTimeout, cacheDuration, statsProvider, new MemoryCache(new MemoryCacheOptions())) {}
+
+        /// <summary>
+        /// Used for injecting an IMemoryCache for unit testing purposes.
+        /// </summary>
+        /// <param name="innerHandler">The inner handler to retrieve the content from on cache misses.</param>
+        /// <param name="maxTimeout">The maximum timeout to wait for the service to respond.</param>
+        /// <param name="cacheDuration">The maximum time span the item should be remained in the cache if not renewed before then.</param>
+        /// <param name="statsProvider">An <see cref="IStatsProvider"/> that records statistic information about the caching behavior.</param>
+        /// <param name="cache">The cache to be used.</param>
+        internal InMemoryCacheFallbackHandler(HttpMessageHandler innerHandler, TimeSpan maxTimeout, TimeSpan cacheDuration, IStatsProvider statsProvider, IMemoryCache cache) : base(innerHandler ?? new HttpClientHandler())
+        {
+            this.StatsProvider = statsProvider ?? new StatsProvider(nameof(InMemoryCacheHandler));
+            this.maxTimeout = maxTimeout;
+            this.cacheDuration = cacheDuration;
+            responseCache = cache ?? new MemoryCache(new MemoryCacheOptions());
+        }
+
+        /// <summary>
+        /// Tries to get the value from the cache, and only calls the delegating handler on cache misses.
+        /// </summary>
+        /// <returns>The HttpResponseMessage from cache, or a newly invoked one.</returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            // only handle GET methods
+            if (request.Method != HttpMethod.Get)
+            {
+                return await base.SendAsync(request, cancellationToken);
+            }
+
+            var key = request.RequestUri.ToString();
+
+            // start 3 tasks
+            var httpSendTask = base.SendAsync(request, cancellationToken);
+            var timeoutTask = Task.Delay(maxTimeout, cancellationToken);
+            var cacheTask = responseCache.TryGetAsync(key);
+            
+            // ensure the send task completes (or the timeout task, whatever comes first).
+            var firstCompletedTask = await Task.WhenAny(httpSendTask, timeoutTask);
+
+            // timeout occurred
+            if (firstCompletedTask == timeoutTask)
```

Comment:
`Task.WhenAny` returns the first completed task, and this check is only true if the timeout task completed before the http send task.